coll/acoll: Add support for MPI_Alltoall()

MithunMohanKadavil opened the pull request:
This PR adds support for MPI_Alltoall() in acoll collective component.
 
For messages lower than a few KBs, the algorithm operates by dividing the n (n=comm size) ranks into f groups (based on the value of rank % f), performing alltoall within the f groups in parallel, following which data is exchanged between groups of f adjacent ranks (starting from rank 0). For example, if f=2 , this algorithm splits the n ranks into 2 groups, one containing all even ranked (rank%2 = 0) processes and another containing all odd ranked (rank%2 = 1) processes. After alltoall is done within these 2 groups (in parallel), adjacent even-odd pairs (pairs being [0,1], [2,3]...) exchange  data to complete MPI_Alltoall  operation. If f=4 or f=8,  alltoall is performed in parallel for 4 or 8 groups respectively, followed by data exchange among 4 or 8 adjacent ranks.
The below diagram captures this algorithm for the case where f=2 and n=8:
![image](https://github.com/user-attachments/assets/766f5c42-5c7c-4537-a376-e377bcfb41c2)
For larger message size range, direct xpmem based copy is used in a linear fashion across all ranks.

The below graphs show the variation in latencies with osu-micro-benchmarks-7.3 for 96 and 192 ranks for tuned and acoll: 
![image](https://github.com/user-attachments/assets/d437f89a-be50-42f3-a7c1-fb369e960632)
![image](https://github.com/user-attachments/assets/4855c73e-b709-4d46-9075-cbfe005df28d)

lrbison reviewed (commented):
My biggest concern is that I don't think this will be correct for derived data types, and I didn't see a catch to check if the buffers are on a GPU (maybe that's at the acoll module level?)

I struggled with these issues myself in my alltoallv implementation, and made what I consider to be a very exhaustive validation tester.  It was trivial to extend it from alltoallv to also test alltoall.  Would you mind giving it a try?  https://github.com/open-mpi/ompi-tests-public/pull/32

Let me also do some testing, in particular I'm curious what would happen on a graviton layout, that can have as much as 96 cores with the same L3 distance.

lrbison commented:
I tried my validator as follows and I see a validation error:

```
mpirun --mca coll_acoll_priority 100 -n 4 ./src/alltoallv_ddt -A alltoall -v 2
```

Looks like I can trigger it on specific tests including `--only 2,2`, `--only 2,3`, and `--only 2,4` all showing failures.

If you dig into the verbose prints you can see:

```
--- Starting test 2,2.  Crossing 0 x 1
Created span from 0:4.  Data from 0:4
Created span from 0:48.  Data from 0:48
Datatype      (send,recv) extents (4,48), size (4,48), and lb (0,0)
Datatype TRUE (send,recv) extents (4,48), size (4,48), and lb (0,0)
<nasty assertion fail abort ... >
```
The "Crossing 0 x 1" refers to send type and receive type respectively in [this switch statement](https://github.com/open-mpi/ompi-tests-public/blob/master/alltoallv_validation/src/alltoallv_ddt.cpp#L194)
So test 2,2 sends 12 4-byte `MPI_INT`s, and receives 1 DDT which is a MPI_Type_contiguous vector of 12 integers.  The following tests make more and more strange DDTs for checking.

MithunMohanKadavil commented:
Hi @lrbison, thanks for sharing the tests. We are testing/validating this code using the tests shared and will soon raise an update fixing the issues/addressing the comments.

bosilca commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+    void *sbuf_vaddr[1] = {(void*)sbuf};
+    void *rbuf_vaddr[1] = {rbuf};
+
+    /* Share this rank's sbuf virtual address with the other ranks,
+     * and similary receive the other rank's sbuf virtual addresses. */
+    error = comm->c_coll->coll_allgather(sbuf_vaddr, sizeof(void *), MPI_BYTE,
+                               data->allshm_sbuf, sizeof(void *), MPI_BYTE,
+                               comm, comm->c_coll->coll_allgather_module);
+    if (MPI_SUCCESS != error) { return error; }
+
+    /* Share this rank's rbuf virtual address with the other ranks,
+     * and similary receive the other rank's rbuf virtual addresses. */
+    error = comm->c_coll->coll_allgather(rbuf_vaddr, sizeof(void *), MPI_BYTE,
+                               data->allshm_rbuf, sizeof(void *), MPI_BYTE,
+                               comm, comm->c_coll->coll_allgather_module);
+    if (MPI_SUCCESS != error) { return error; }
```
No doubts about this. However, I question the need for the second allgather, it seems to exchange the receive buffer but the next step of the algorithm does a pull from remote memory, so knowing their receive buffer is irrelevant. Am I misreading the code ?

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+            (*sync_enable) = false;
+            (*split_factor) = 2;
+        } else {
+            (*sync_enable) = true;
```
On our hardware, we observed the linear variants to perform better for lower node counts. However for higher node counts, it still needs to be tuned, and at that point this decision function will be updated.

lrbison replied:
@MithunMohanKadavil Rather than requiring a code-update to update your customer's tuning, would you take a look at [this PR ](https://github.com/open-mpi/ompi/pull/13104) and see if it has the required features to support a tuning file format for acoll?  I haven't taken a look at exactly what tuning tree you use, but if you could sketch out how you would lay out the file structure, maybe we can make it compatible with han and move towards a generic tuning file format with at least the outer structure shared across multiple collective components.

MithunMohanKadavil replied:
Hi @lrbison , the framework in the PR does have the features required to support the tuning file. However we are yet to decide on the file structure, in the sense that what parameters will be tunable and how it can be applied across the algorithms in acoll. Additionally more tunable parameters may come up for us and as such the tuning file structure still needs to be vetted out from our side.

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+    }
+
+error_handler:
+    ;
```
This was done to avoid having multiple return statements in the function. However since there was no cleanup involved, had to leave it empty.

MithunMohanKadavil replied:
Updated

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+    }
+}
+
+static size_t mca_coll_acoll_get_msg_thres(coll_acoll_subcomms_t *subc)
```
Updated.

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+
+static size_t mca_coll_acoll_get_msg_thres(coll_acoll_subcomms_t *subc)
+{
+    size_t msg_thres[DIST_END] = {4096, 2048, 1024, 1024, 512};
```
Added new mca parameter "coll_acoll_alltoall_psplit_msg_thresh" to tune the same.

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+
+static size_t mca_coll_acoll_get_xpmem_switch_msg_thres(coll_acoll_subcomms_t *subc)
+{
+    size_t msg_thres[DIST_END] = {512, 256, 128, 128, 128};
```
This function has been removed since the xpmem based linear alltoall path (for which this was relevant) has been removed.

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+        memcpy((char*)rbuf + write_pos, (char*)data->xpmem_saddr[next_rank] + read_pos,
+            rcount * rext);
+    }
+    error = ompi_datatype_copy_content_same_ddt(rdtype, rcount,
```
This xpmem based function (containing this copy) has been removed for now due to issues arising from using derived datatypes.

MithunMohanKadavil commented:
bot:retest

MithunMohanKadavil commented on ompi/mca/coll/acoll/coll_acoll_alltoall.c:
```diff
@@ -0,0 +1,618 @@
+
+        /* Create a new datatype that iterates over the send buffer in strides
+         * of ps_grp_size * scount. */
+        struct ompi_datatype_t *new_ddt;
```
Hi @bosilca , yeah the current approach i have used is sub optimal. The only change across types is the dipls array, based on the read_pos. I am not aware of how the displs is mapped to the lower bound and it would be very helpful if you could share an example or API that can be used to manipulate the lower bound of an already created data type.

As for point 2, if the comm_size is a perfect multiple of split factor, then each rank would contain data for split factor numbers of adjacent ranks (including its own) times ps_grp_num_ranks. However in case comm_size is not a multiple of split factor, then the original split factor itself would have been modified such that comm_size is expressed as (new_comm_size * split_factor) + 1, and the data for the final rank will be handled in the function mca_coll_acoll_last_rank_scatter_gather.

bosilca replied:
`ompi_datatype_create_resized` is what you're looking for.

MithunMohanKadavil replied:
Thanks, will check this and rework the code.

MithunMohanKadavil replied:
Hi @bosilca , i have reworked the code to avoid datatype creation inside the loop in mca_coll_acoll_exchange_data function. Additionally i am creating a vector datatype with non unit strides to facilitate the same functionality as before. And the start point in the send buffer is adjusted, so lower bounds also need not be updated. Thanks for the suggestions regarding the datatype usage improvements.

jsquyres commented:
@MithunMohanKadavil Can you rebase instead of merge?  We tend to prefer that.  Thanks!

MithunMohanKadavil commented:
> @MithunMohanKadavil Can you rebase instead of merge? We tend to prefer that. Thanks!

jsquyres commented:
@MithunMohanKadavil No worries.  No need to close this PR -- you can just remove the merge commit and rebase.

MithunMohanKadavil commented:
@jsquyres Remove merge commit and updated the PR, thanks.

jsquyres commented:
> @jsquyres Removed merge commit and updated the PR, thanks.

Thank you!
